Markdown export

uroybd opened the pull request:
This PR enables users to export markdown locally with some configuration options to allow users to format the output to a certain extent.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/9076)
<!-- Reviewable:end -->

zwim commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,163 @@
+                end,
+            },
+            {
+                text = _("underscore"),
```
Maybe uppercase "Underscore"

uroybd replied:
FIxed.

zwim commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,163 @@
+    extension = "md",
+}
+
+
```
Maybe one empty line

uroybd replied:
FIxed.

zwim commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,163 @@
+    self.settings = plugin_settings[self.name] or {}
+end
+
+
```
ditto

uroybd replied:
FIxed.

zwim commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,163 @@
+local _ = require("gettext")
+local UIManager = require("ui/uimanager")
+
+
```
Maybe on empty line

uroybd replied:
FIxed.

zwim commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,163 @@
+
+function MarkdownExporter:editFormatStyle(drawer_style)
+    local radio_buttons = {}
+    for _, v in ipairs(formatter_buttons) do
```
With `_` here you overwrite `gettext`

uroybd replied:
Yes, but not using `gettext` in the loop.

uroybd replied:
Thank you. FIxed.

zwim reviewed (commented):
Not familiar with markdown exporter, but some eye-catchers mentioned.

zwim commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,160 @@
+    return true
+end
+
+return MarkdownExporter
```
At the end there should be a newline. (Sorry overlooked that first.)

uroybd replied:
You have not. I removed it.

Added it back again.

hius07 commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,160 @@
+        })
+    end
+    UIManager:show(require("ui/widget/radiobuttonwidget"):new{
+        title_text = _("Highlight style for " .. drawer_style),
```
You are putting `drawer_style` into the translation.

uroybd replied:
Moved it out of the transtation. :)

pazos replied:
Please move to a template instead `T(_("Highlight style for %1"), drawer_style)`

uroybd replied:
Done.

hius07 commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,160 @@
+                end,
+            },
+            {
+                text = _("Underscore"),
```
In interface it is "Underline":
https://github.com/koreader/koreader/blob/4b0f7471f4a38c3371d01a8251d8f9f9f2937eb8/frontend/apps/reader/modules/readerhighlight.lua#L304

uroybd replied:
FIxed. Thanks.

hius07 commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,160 @@
+        })
+    end
+    UIManager:show(require("ui/widget/radiobuttonwidget"):new{
+        title_text = _("Formatting style for ")  .. drawer_style,
```
Nit: double space.
Maybe it is better to use labels' text (already translated strings) instead of the internal names for `drawer_style`.

uroybd replied:
Tried to fix this. Is it okay now?

pazos commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,160 @@
+    extension = "md",
+}
+
+function MarkdownExporter:loadSettings()
```
remove. You're using the exact same logic as `BaseExporter`.

Also this function isn't intended to be overriden in subclasses. It should work fine on all plugins. If not we need to refactor the function in the parent class.

uroybd replied:
Ops! Sorry. was toying with that. Forgot to remove.

Removed now.

pazos commented:
@uroybd: Looks good. But you added quite a bit of configuration options and now I'm worried how it will look in the Joplin target.

Could you adapt the Joplin target to inherit from the markdown one as part of this PR?. 

This will show which parts of the markdown plugin need to be more atomic (ie: the settings for markdown highlights "translation" can be moved to its own function that returns a sub_item_table so both targets can share a bit of code, like:

```lua
function JoplinExporter:getMenuTable()
    local t = {
        --- proper joplin settings here
    }
    -- append here common markdown settings
    for i, v in ipairs(self:getMarkdownSettingsMenuTable()) do
        table.insert(t.sub_item_table, #t.sub_item_table, v)
    end
end
```

uroybd commented:
> @uroybd: Looks good. But you added quite a bit of configuration options and now I'm worried how it will look in the Joplin target.
> 
> Could you adapt the Joplin target to inherit from the markdown one as part of this PR?. 
> 
> This will show which parts of the markdown plugin need to be more atomic (ie: the settings for markdown highlights "translation" can be moved to its own function that returns a sub_item_table so both targets can share a bit of code, like:
> 
> ```lua
> function JoplinExporter:getMenuTable()
>     local t = {
>         --- proper joplin settings here
>     }
>     -- append here common markdown settings
>     for i, v in ipairs(self:getMarkdownSettingsMenuTable()) do
>         table.insert(t.sub_item_table, #t.sub_item_table, v)
>     end
> end
> ```
> 

Inheriting code-wise will be easiest part, really.

I'm thinking about inheriting the settings itself. If an user likes, for example invert highlights to be bold italic for local, normally she/he will want the same for joplin too. There's no reason they should set same settings twice. 

That's why I wanted only one new checkbox for joplin to enable markdown formatting (from markdown plugin).

hius07 commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,157 @@
+    UIManager:show(require("ui/widget/radiobuttonwidget"):new{
+        title_text = T(_("Formatting style for %1"), _(label)),
+        width_factor = 0.8,
+        keep_shown_on_apply = true,
```
Is it needed?

uroybd replied:
No. Removed it. :)

hius07 commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,157 @@
+                callback = function() self.settings.highlight_formatting = not self.settings.highlight_formatting end,
+            },
+            {
+                text = _("Lighten"),
```
Maybe it would be useful to show current setting of the format style in the menu item?

uroybd replied:
I thought about that. Can I generate label from a function?

uroybd replied:
Surely I can. Gettext is also a function, right?

uroybd replied:
I've generated labels using functions. However, it is not being updated on settings changes. What am I missing?

hius07 replied:
Update touchmenu instance in callback.

uroybd replied:
Thanks. Done. :D

pazos commented:
> That's why I wanted only one new checkbox for joplin to enable markdown formatting (from markdown plugin).

Then do it as you wish and lets review that.

uroybd commented:
> > That's why I wanted only one new checkbox for joplin to enable markdown formatting (from markdown plugin).
> 
> Then do it as you wish and lets review that.

@pazos 

In the last commit I did the following things:

1. I've moved the markdown formatting function and relevant code to `template/md.lua`.
2. I'm pulling markdown plugin-specific settings from storage in Joplin export and using that to format.
3. Added a Markdown formatting help in Joplin exporter settings directing users to show where to go.
4. For some reason, notebook_guid is not being saved, added a check there which fixed the update problem.

pazos reviewed (changes requested):
Good in general.

Please check what happens if you try to export to joplin without "populating" markdown settings first.

uroybd commented:
> Good in general.
> 
> Please check what happens if you try to export to joplin without "populating" markdown settings first.
> 

I'll check.

If I am not wrong, whenever exporter plugin get initialized, it will call all the menu table creation functions, right? So no need to populate in export too and the settings will always be generated.

pazos commented:
> If I am not wrong, whenever exporter plugin get initialized, it will call all the menu table creation functions, right? So no need to populate in export too and the settings will always be generated.

I don't think so. The only function called on plugin load is the `init` for each exporter, which calls `self:loadSettings`.

Each exporter `getMenuTable` should be called when the FM/Reader touch menu is instanciated or when you enter plugin menus. I don't remember.

uroybd commented:
I need an way to generate the default settings. What should be the preferred way?

pazos commented:
> I need an way to generate the default settings. What should be the preferred way?


```lua
--[[--
Called after loading the exporter and its settings.

You can use it if the exporter requires some tweaks before it is able to show their own UI settings
]]
function BaseExporter:onInit()
end
```

```lua
function MarkdownExporter:onInit()
    if not self.settings.formatting_options or not self.settings.highlight_formatting then
        self.settings.formatting_options = self.settings.formatting_options or {
            lighten = "italic",
            underscore = "underline_markdownit",
            strikeout = "strikethrough",
            invert = "bold",
        }
        self.settings.highlight_formatting = self.settings.highlight_formatting or true
        self:saveSettings()
    end
end
```


add `self:onInit()` after `self:loadSettings` in `BaseExporter:_init()` 

Remove the function `MarkdownExporter:populateSettings()` and all its invokations.

uroybd commented:
@pazos

`onInit` invoked in `_init` always uses the base version of the exporter (so does `loadSettings`) I called it in the `main.lua` where I'm setting the path to the plugins in a loop.

hius07 commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,142 @@
+            {
+                text = _("Format Highlights based on style"),
+                checked_func = function() return self.settings.highlight_formatting end,
+                callback = function() self.settings.highlight_formatting = not self.settings.highlight_formatting end,
```
Next 4 style menu items might be disabled depending on this setting.

Some code duplication might be avoid via cycling through something like
https://github.com/koreader/koreader/blob/60b043ebe9f0dd82b102bd6c9db92c6eab52d1d4/frontend/apps/reader/modules/readerhighlight.lua#L302-L307

uroybd replied:
Done.

hius07 commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,129 @@
+
+function MarkdownExporter:getFormatterLabel(header, drawer_style)
+    return T("%1: %2", header, md.formatters[self.settings.formatting_options[drawer_style]].label)
+end
```
Might be done in the cycle, no need for the function.

uroybd replied:
Done.

uroybd commented:
@pazos 
I'm explicitly checking if `highlight_formatting` is `nil` (not falsy) because `false` is an acceptable value where nil is an absence of that value.

hius07 commented:
Thanks, looks good from my side.

pazos commented:
> Seems like this PR is in good shape now. 

It is, indeed.

There are some issues left, some of them inherited. We can tackle a few of them here if you want.


--------------------------------
1. joplin help instructions are a bit too much to be included in plugin "help" infomessage.
2. joplin markdown help instructions sound redundant with the other help

*Proposal:* 

1. Create a new wiki page. Move technical info about how to setup the computer part (and to add a token easily) there.
2. Join the two "helps" into a single one. Which contains the info included in the markdown help followed by "To setup joplin sync please go to https://github.com/koreader/wiki/new_page.


--------------------------------
1. joplin changed its output, now it is markdown.

*Proposal:* 

1. Figure out if it needs a version bump or if the version bump is just for other formats (mostly json ATM) and the rest of targets should keep its current "target/1.0.0" version, no matter if the output changed.


> We will wait for 2022.05 to be released before this will be merged, right?

Not needed, IMHO. These are new addons but don't change the workflow enough to justify waiting more time.

uroybd commented:
@pazos 

For "help", I thought that too. Let's merge them into one and move the setup details to a wiki page.

I think, we should make a version bump for Joplin. The data representation has changed. Markdown is primarily a 'visual' format, but also a pretty parsable data format. Bumping version and adding that info at the bottom like:

_Generated on: <time>, Version: 2.0.0_

This is a minor thing, let me know what you would like.

uroybd commented:
@pazos @Frenzie 
I thought there was a moderation in the Wiki. That's why I edited the doc here: https://github.com/koreader/koreader/wiki/Highlight-export

and added a note explaining why I did that. I'm sorry. Should I revert back?

uroybd commented:
> @pazos @Frenzie I thought there was a moderation in the Wiki. That's why I edited the doc here: https://github.com/koreader/koreader/wiki/Highlight-export
> 
> and added a note explaining why I did that. I'm sorry. Should I revert back?

With this changes, we can replace the current help text for Joplin to something like this as @pazos suggested:

You can configure markdown formatting by navigating to Export Highlights > Choose Formats and Services > Markdown.

For joplin setup instructions please visit https://github.com/koreader/koreader/wiki/Highlight-export#joplin.

uroybd commented:
@pazos 

Hi, you're probably very busy. Let me know what you think about this once you're free?

uroybd commented:
@pazos and @Frenzie

Now that 2022.5 has been released, can you please take a look into it?

Frenzie commented:
Anything in particular? Some screenshots might help. :-)

uroybd commented:
> > Seems like this PR is in good shape now. 
> 
> It is, indeed.
> 
> There are some issues left, some of them inherited. We can tackle a few of them here if you want.
> 
> 
> --------------------------------
> 1. joplin help instructions are a bit too much to be included in plugin "help" infomessage.
> 2. joplin markdown help instructions sound redundant with the other help
> 
> *Proposal:* 
> 
> 1. Create a new wiki page. Move technical info about how to setup the computer part (and to add a token easily) there.
> 2. Join the two "helps" into a single one. Which contains the info included in the markdown help followed by "To setup joplin sync please go to https://github.com/koreader/wiki/new_page.
> 
> 
> --------------------------------
> 1. joplin changed its output, now it is markdown.
> 
> *Proposal:* 
> 
> 1. Figure out if it needs a version bump or if the version bump is just for other formats (mostly json ATM) and the rest of targets should keep its current "target/1.0.0" version, no matter if the output changed.
> 
> 
> > We will wait for 2022.05 to be released before this will be merged, right?
> 
> Not needed, IMHO. These are new addons but don't change the workflow enough to justify waiting more time.
> 
> 

@Frenzie 

Referring to this actually.

I've updated the doc. Is it okay?  I have to know that should we update version number for joplin too.

Frenzie commented on plugins/exporter.koplugin/target/joplin.lua:
```diff
@@ -305,16 +287,10 @@ function JoplinExporter:getMenuTable()
-To export to Joplin, you must forward the IP and port used by this plugin to the localhost:port on which Joplin is listening. This can be done with socat or a similar program. For example:
-
-For Windows: netsh interface portproxy add v4tov4 listenaddress=0.0.0.0 listenport=41185 connectaddress=localhost connectport=41184
+                        text = T(_([[You can configure markdown formatting by navigating to:
+        Export Highlights > Choose Formats and Services > Markdown
```
```suggestion
                        text = T(_([[Markdown formatting can be configured in:
        Export highlights > Choose formats and services > Markdown
```
(if that says Export Highlights and Choose Formats and Services, please remove those capitals there)

Frenzie commented on plugins/exporter.koplugin/target/joplin.lua:
```diff
@@ -305,16 +287,10 @@ function JoplinExporter:getMenuTable()
-
-For more information, please visit https://github.com/koreader/koreader/wiki/Highlight-export.]])
-                            , BD.dirpath("example"))
+For joplin setup instructions please visit https://github.com/koreader/koreader/wiki/Highlight-export#joplin.]]))
```
```suggestion
For Joplin setup instructions, see https://github.com/koreader/koreader/wiki/Highlight-export#joplin.]]))
```

Frenzie commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,117 @@
+
```
random newline

Frenzie commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,117 @@
+
+local md = require("template/md")
+local UIManager = require("ui/uimanager")
+local util = require("ffi/util")
+local T = util.template
+local _ = require("gettext")
```
order

uroybd replied:
`T` depends on `util` so ordered it like this:
```lua
local md = require("template/md")
local util = require("ffi/util")
local UIManager = require("ui/uimanager")
local T = util.template
local _ = require("gettext")
```

Is it okay now?

Frenzie replied:
My bad, I should've linked to the style guide: https://github.com/koreader/koreader-base/wiki/Coding-style#order-of-requires

uroybd replied:
Is it fine now?

Frenzie replied:
Yup, thanks!

Frenzie commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,117 @@
+local formatter_buttons = {
+    {_("None"), "none"},
+    {_("Bold"), "bold"},
+    {_("Bold Italic"), "bold_italic"},
```
```suggestion
    {_("Bold italic"), "bold_italic"},
```

Frenzie commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,117 @@
+                callback = function() self:toggleEnabled() end,
+            },
+            {
+                text = _("Format Highlights based on style"),
```
```suggestion
                text = _("Format highlights based on style"),
```

Frenzie commented on plugins/exporter.koplugin/template/md.lua:
```diff
@@ -0,0 +1,62 @@
+    },
+    bold_italic = {
+        formatter = "**_%s_**",
+        label = _("Bold Italic")
```
```suggestion
        label = _("Bold italic")
```

uroybd commented:
@Frenzie 
With these drastic changes in Joplin output, I believe @pazos won't mind my version bumping Joplin Exporter. Besides this, he was satisfied with this PR. You might consider merging this if you don't have any more feedback.

pazos commented:
Sorry for the lack of response. Will review soon.

In the mean time @uroybd might want to look at https://www.mobileread.com/forums/showthread.php?t=346728, it seems that we have our first bug reports :1st_place_medal: 

Did you test the refactor on a real device?. I did not :p.

pazos commented on plugins/exporter.koplugin/target/joplin.lua:
```diff
@@ -15,6 +15,7 @@ local JoplinExporter = require("base"):new {
     name = "joplin",
     is_remote = true,
     notebook_name = _("KOReader Notes"),
+    version = "2.0.0",
```
"1.0.1" or "1.1.0" as you wish. Definitly not "2.0.0"

uroybd replied:
1.1.0 Since changes are not patch level.

uroybd commented:
> Sorry for the lack of response. Will review soon.
> 
> In the mean time @uroybd might want to look at https://www.mobileread.com/forums/showthread.php?t=346728, it seems that we have our first bug reports 1st_place_medal
> 
> Did you test the refactor on a real device?. I did not :p.

Interesting. I've just checked the exporting with the steps described in that post. However, Likebook P78 produced the exports perfectly. Unfortunately, I don't have any other e-readers. :(

uroybd commented:
> > Sorry for the lack of response. Will review soon.
> > In the mean time @uroybd might want to look at https://www.mobileread.com/forums/showthread.php?t=346728, it seems that we have our first bug reports 1st_place_medal
> > Did you test the refactor on a real device?. I did not :p.
> 
> Interesting. I've just checked the exporting with the steps described in that post. However, Likebook P78 produced the exports perfectly. Unfortunately, I don't have any other e-readers. :(

@pazos

Jailbroken a kindle PW4 of one of my coworkers to test this. Apparently, It's the colons in the filename. Fixed it.

Frenzie commented:
@uroybd Please use http://koreader.rocks/doc/modules/util.html#getSafeFilename instead of doing manual replacements.

uroybd commented:
> @uroybd Please use http://koreader.rocks/doc/modules/util.html#getSafeFilename instead of doing manual replacements.

Done!

Frenzie commented:
I'll wait for @pazos to take a final look. ^_^

pazos commented on plugins/exporter.koplugin/target/joplin.lua:
```diff
@@ -305,16 +288,10 @@ function JoplinExporter:getMenuTable()
-
-For more information, please visit https://github.com/koreader/koreader/wiki/Highlight-export.]])
-                            , BD.dirpath("example"))
+For Joplin setup instructions, see https://github.com/koreader/koreader/wiki/Highlight-export#joplin.]]))
```
The link shouldn't be translatable

Also the link looks a bit long. It would be better to have a link `https://github.com/koreader/koreader/wiki/joplin` here (that can be linked in general `https://github.com/koreader/koreader/wiki/Highlight-export`)

uroybd replied:
Done.

uroybd commented:
@pazos 

Is anything left to do in this PR?

pazos reviewed (approved):
lgtm

Frenzie commented on plugins/exporter.koplugin/target/markdown.lua:
```diff
@@ -0,0 +1,135 @@
+local formatter_buttons = {
+    { _("None"), "none" },
+    { _("Bold"), "bold" },
+    { _("Bold Italic"), "bold_italic" },
```
```suggestion
    { _("Bold italic"), "bold_italic" },
```
